cmd/query/app/querysvc/adjuster/spaniduniquifier_test.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2019 The Jaeger Authors.
+// Copyright (c) 2017 Uber Technologies, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package adjuster
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/internal/jptrace"
+)
+
+var (
+	clientSpanID  = pcommon.SpanID([]byte{0, 0, 0, 0, 0, 0, 0, 1})
+	anotherSpanID = pcommon.SpanID([]byte{1, 0, 0, 0, 0, 0, 0, 0})
+)
+
+func makeTraces() ptrace.Traces {
+	traceID := pcommon.TraceID([]byte{0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 3})
+
+	traces := ptrace.NewTraces()
+	spans := traces.ResourceSpans().AppendEmpty().ScopeSpans().AppendEmpty().Spans()
+
+	clientSpan := spans.AppendEmpty()
+	clientSpan.SetTraceID(traceID)
+	clientSpan.SetSpanID(clientSpanID)
+	clientSpan.SetKind(ptrace.SpanKindClient)
+
+	serverSpan := spans.AppendEmpty()
+	serverSpan.SetTraceID(traceID)
+	serverSpan.SetSpanID(clientSpanID) // shared span ID
+	serverSpan.SetKind(ptrace.SpanKindServer)
+
+	anotherServerSpan := spans.AppendEmpty()
+	anotherServerSpan.SetTraceID(traceID)
+	anotherServerSpan.SetSpanID(clientSpanID) // shared span ID
+	anotherServerSpan.SetKind(ptrace.SpanKindServer)
+
+	anotherSpan := spans.AppendEmpty()
+	anotherSpan.SetTraceID(traceID)
+	anotherSpan.SetSpanID(anotherSpanID)
+
+	return traces
+}
+
+func TestSpanIDUniquifierTriggered(t *testing.T) {
+	traces := makeTraces()
+	deduper := SpanIDUniquifier()
+	require.NoError(t, deduper.Adjust(traces))
+
+	spans := traces.ResourceSpans().At(0).ScopeSpans().At(0).Spans()
+
+	clientSpan := spans.At(0)
+	assert.Equal(t, clientSpanID, clientSpan.SpanID(), "client span ID should not change")
+
+	serverSpan := spans.At(1)
+	assert.EqualValues(t, []byte{0, 0, 0, 0, 0, 0, 0, 2}, serverSpan.SpanID(), "server span ID should be reassigned")
+	assert.EqualValues(t, []byte{0, 0, 0, 0, 0, 0, 0, 3}, serverSpan.ParentSpanID(), "next server span should be this server span's parent")
+
+	anotherServerSpan := spans.At(2)
+	assert.EqualValues(t, []byte{0, 0, 0, 0, 0, 0, 0, 3}, anotherServerSpan.SpanID(), "server span ID should be reassigned")
+	assert.EqualValues(t, clientSpanID, anotherServerSpan.ParentSpanID(), "client span should be server span's parent")
```

Comment:
@yurishkuro does this expectation make sense to you?

Reply:
first, when you have one `client` and two `server` spans sharing the same ID, it's doesn't really make sense and the behavior of adjuster is not well defined, because the condition already violates "shared span id" notion from Zipkin. I think it's technically possible to have if a retry from client was done without creating a new client span. I think a reasonable thing to do in such case is to assign unique IDs to both `server` spans, such that we get a tree (but not a chain that you have):
```
clientSpan -> serverSpan
clientSpan -> anotherServerSpan
``
In addition, if there is a 4th span whose `parentId == clientSpanId`, then that 4th span's parent ID must be overwritten to the newly generated span ID. Here we have an ambiguity - it could be either `serverSpan.spanId` or `anotherServerSpan.spanId`, there's no way for us to know (and we should arbitrarily pick first or last reassigned `server` span).

Reply:
@yurishkuro Ah okay. I'm happy to remove this test if the behaviour is undefined then because it didn't exist in the old implementation either.